graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlErrorsHandlers.java

```diff
@@ -0,0 +1,106 @@
+ */
+
+package com.linecorp.armeria.server.graphql;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+
+import graphql.ExecutionResult;
+import graphql.ExecutionResultImpl;
+import graphql.GraphQLError;
+import graphql.GraphqlErrorException;
+import graphql.validation.ValidationError;
+
+final class GraphqlErrorsHandlers {
+
+    private static final Logger logger = LoggerFactory.getLogger(GraphqlErrorsHandlers.class);
+
+    /**
+     * Returns a {@link HttpResponse} based on {@link Throwable} or List of {@link GraphQLError}.
+     * @param errorsMappingFunction The function which maps the {@link GraphQLError} to an {@link HttpStatus}
+     */
+    static GraphqlErrorsHandler of(GraphqlErrorsMappingFunction errorsMappingFunction) {
+        return (ctx, input, produceType, executionResult, cause) -> {
+            if (cause != null) {
+                // graphQL.executeAsync() returns an error in the executionResult with getErrors().
+                // Use 500 Internal Server Error because this cause might be unexpected.
+                final ExecutionResult error = newExecutionResult(cause);
+                return HttpResponse.ofJson(HttpStatus.INTERNAL_SERVER_ERROR, produceType,
+                                           error.toSpecification());
+            }
+            final List<GraphQLError> errors = executionResult.getErrors();
+            final HttpStatus httpStatus = errorsMappingFunction.orElse(GraphqlErrorsMappingFunction.of())
+                                                               .apply(ctx, input, errors);
+            return toHttpResponse(requireNonNull(httpStatus, "httpStatus"), executionResult, produceType);
+        };
+    }
+
+    private GraphqlErrorsHandlers() {}
+
+    /**
+     * Return {@link ExecutionResult} based {@link Throwable}.
+     */
+    static ExecutionResult newExecutionResult(Throwable cause) {
+        return new ExecutionResultImpl(GraphqlErrorException.newErrorException()
+                                                            .message(cause.getMessage())
+                                                            .cause(cause)
+                                                            .build());
+    }
+
+    /**
+     * Return {@link HttpStatus} based List of {@link GraphQLError}.
+     */
+    static HttpStatus graphqlErrorsToHttpStatus(List<GraphQLError> errors) {
+        if (errors.isEmpty()) {
+            return HttpStatus.OK;
+        }
+        if (errors.stream().anyMatch(ValidationError.class::isInstance)) {
+            // The server SHOULD deny execution with a status code of 400 Bad Request for
+            // invalidate documentation.
+            return HttpStatus.BAD_REQUEST;
+        }
+        return HttpStatus.UNKNOWN;
```

Comment:
`UNKNOWN` is for internal use only so we should use `SERVICE_UNAVAILABLE` instead.
How about merging `GraphqlErrorsHandler` and `GraphqlErrorsMappingFunction` into `GraphqlErrorsHandler`?
I don't see any particular reason to separate them for now. If we really need the mapping function, we can add it later.

Reply:
I didn't know about `UNKNOWN` and `SERVICE_UNAVAILABLE`, because I haven't written the Http status code in detail yet. I'll fix it right away. Thank you!

The reason for separating Handler and MappingFunction is that I wanted to provide a simple way to convert GraphQL error to Http response. And shamefully, I couldn't think of implementing the orElse method on Handler, so I decided that Handler should have only one processing logic.

Additionally, if MappingFunction is merged with Handler, I would like to reduce the number of Parameters of Handler, is there a good way?

If only `ServiceRequestContext`, `ExecutionInput`, and `List<GraphQLError>` were delivered as you first restricted, the logic already processed in the `execute` would be re-executed in the Handler, so I thought the logic was inefficient.

What are the challenges of implementing the Handler interface as this?
```java
public interface GraphqlErrorsHandler {
    handler(ExecutionResult executionResult, Throwable cause, List<GraphQLError> errors) {
        ...
    }
}
```

Reply:
Adding `ServiceRequestContext` and `ExecutionInput` to the parameter is not for executing the method twice, but for providing more information to the users who customize the functions so that they can use additional information while creating an `HttpResponse`.

> What are the challenges of implementing the Handler interface as this?

How about adding `ServiceRequestContext` and `ExecutionInput` as our current interface does?
```java
public interface GraphqlErrorsHandler {
    handle(ServiceRequestContext, ExecutionInput input, ExecutionResult result, MediaType negotiatedProduceType, @Nullable Throwable cause) {
        ...
    }
}
```
Please note that we don't need to pass `List<GraphQLError>` because we can get it from the `ExecutionResult`.

Reply:
Thank you for your kind advice. I solved the problem related to Http status, and as you advised, I merged MappingFunction into ErrorsHandler. After modifying the code, I think this is much better, too.